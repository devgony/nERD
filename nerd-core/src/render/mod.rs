use crate::models::{Entity, Schema, Relationship};
use ratatui::{
    layout::{Constraint, Direction, Layout, Rect},
    style::{Color, Modifier, Style},
    text::Line,
    widgets::{Block, Borders, Clear, List, ListItem, Paragraph},
    Frame,
};

pub struct DiagramRenderer {
    canvas_width: u16,
    canvas_height: u16,
}

impl DiagramRenderer {
    pub fn new(width: u16, height: u16) -> Self {
        Self {
            canvas_width: width,
            canvas_height: height,
        }
    }

    pub fn render(&self, f: &mut Frame, schema: &Schema, area: Rect, selected_entity: &Option<String>) {
        if schema.entities.is_empty() {
            self.render_empty_state(f, area);
            return;
        }

        self.render_entities(f, schema, area, selected_entity);
        self.render_relationships(f, schema, area);
    }

    fn render_empty_state(&self, f: &mut Frame, area: Rect) {
        let empty_message = Paragraph::new("Loading sample schema... Press 's' to view SQL or 'r' to refresh layout.")
            .style(Style::default().fg(Color::DarkGray))
            .block(Block::default().borders(Borders::ALL).title("ERD Diagram"));
        f.render_widget(empty_message, area);
    }

    fn render_entities(
        &self,
        f: &mut Frame,
        schema: &Schema,
        area: Rect,
        selected_entity: &Option<String>,
    ) {
        for (entity_name, entity) in &schema.entities {
            let entity_area = self.calculate_entity_area(entity, area);
            let is_selected = selected_entity.as_ref() == Some(entity_name);
            
            self.render_single_entity(f, entity, entity_area, is_selected);
        }
    }

    fn render_single_entity(
        &self,
        f: &mut Frame,
        entity: &Entity,
        area: Rect,
        is_selected: bool,
    ) {
        let border_style = if is_selected {
            Style::default().fg(Color::Yellow).add_modifier(Modifier::BOLD)
        } else {
            Style::default().fg(Color::White)
        };

        let title_style = if is_selected {
            Style::default().fg(Color::Yellow).add_modifier(Modifier::BOLD)
        } else {
            Style::default().fg(Color::Cyan).add_modifier(Modifier::BOLD)
        };

        let block = Block::default()
            .title(entity.name.clone())
            .title_style(title_style)
            .borders(Borders::ALL)
            .border_style(border_style);

        let inner_area = block.inner(area);

        let columns: Vec<ListItem> = entity
            .columns
            .iter()
            .map(|col| {
                let key_indicator = if col.is_primary_key {
                    "🗝 "
                } else if col.is_foreign_key {
                    "🔗 "
                } else {
                    "  "
                };

                let nullable_indicator = if col.nullable { "?" } else { "" };

                let column_text = format!(
                    "{}{} {}{}",
                    key_indicator,
                    col.name,
                    col.data_type,
                    nullable_indicator
                );

                let style = if col.is_primary_key {
                    Style::default().fg(Color::Yellow).add_modifier(Modifier::BOLD)
                } else if col.is_foreign_key {
                    Style::default().fg(Color::Magenta)
                } else {
                    Style::default().fg(Color::White)
                };

                ListItem::new(column_text).style(style)
            })
            .collect();

        let column_list = List::new(columns).block(Block::default());

        f.render_widget(block, area);
        f.render_widget(column_list, inner_area);
    }

    fn render_relationships(
        &self,
        f: &mut Frame,
        schema: &Schema,
        area: Rect,
    ) {
        for relationship in &schema.relationships {
            if let (Some(from_entity), Some(to_entity)) = (
                schema.entities.get(&relationship.from_table),
                schema.entities.get(&relationship.to_table),
            ) {
                self.render_relationship_line(f, from_entity, to_entity, relationship, area);
            }
        }
    }

    fn render_relationship_line(
        &self,
        f: &mut Frame,
        from_entity: &Entity,
        to_entity: &Entity,
        relationship: &Relationship,
        area: Rect,
    ) {
        let from_area = self.calculate_entity_area(from_entity, area);
        let to_area = self.calculate_entity_area(to_entity, area);

        // Calculate connection points for specific columns
        let (from_point, to_point) = self.calculate_column_connection_points(
            from_entity, 
            to_entity, 
            &relationship.from_column,
            &relationship.to_column,
            &from_area, 
            &to_area
        );

        self.draw_connection_line(f, from_point, to_point, area);
    }


    fn calculate_column_connection_points(
        &self,
        from_entity: &Entity,
        to_entity: &Entity,
        from_column_name: &str,
        to_column_name: &str,
        from_area: &Rect,
        to_area: &Rect,
    ) -> ((u16, u16), (u16, u16)) {
        // Calculate the Y position of the specific columns within their entities
        let from_column_y = self.calculate_column_y_position(from_entity, from_column_name, from_area);
        let to_column_y = self.calculate_column_y_position(to_entity, to_column_name, to_area);

        // Determine which edges to connect based on entity positions
        let from_center_x = from_area.x + from_area.width / 2;
        let from_center_y = from_area.y + from_area.height / 2;
        let to_center_x = to_area.x + to_area.width / 2;
        let to_center_y = to_area.y + to_area.height / 2;

        let from_point = if to_center_x > from_center_x {
            // Connect from right edge at column height
            (from_area.x + from_area.width - 1, from_column_y)
        } else if to_center_x < from_center_x {
            // Connect from left edge at column height
            (from_area.x, from_column_y)
        } else if to_center_y > from_center_y {
            // Connect from bottom edge
            (from_center_x, from_area.y + from_area.height - 1)
        } else {
            // Connect from top edge
            (from_center_x, from_area.y)
        };

        let to_point = if from_center_x > to_center_x {
            // Connect to right edge at column height
            (to_area.x + to_area.width - 1, to_column_y)
        } else if from_center_x < to_center_x {
            // Connect to left edge at column height
            (to_area.x, to_column_y)
        } else if from_center_y > to_center_y {
            // Connect to bottom edge
            (to_center_x, to_area.y + to_area.height - 1)
        } else {
            // Connect to top edge
            (to_center_x, to_area.y)
        };

        (from_point, to_point)
    }

    fn calculate_column_y_position(&self, entity: &Entity, column_name: &str, entity_area: &Rect) -> u16 {
        // The inner area starts 1 pixel down from the entity area (for the border)
        let inner_y = entity_area.y + 1;
        
        // Find the column index
        let column_index = entity.columns
            .iter()
            .position(|col| col.name == column_name)
            .unwrap_or(0);
        
        // Each column takes up 1 line in the list
        // Add the column index to the inner Y position
        inner_y + column_index as u16
    }

    fn draw_connection_line(
        &self,
        f: &mut Frame,
        from: (u16, u16),
        to: (u16, u16),
        area: Rect,
    ) {
        // Draw a proper line between the two points
        self.draw_line_between_points(f, from, to, area);
    }

    fn draw_line_between_points(
        &self,
        f: &mut Frame,
        from: (u16, u16),
        to: (u16, u16),
        area: Rect,
    ) {
        let dx = to.0 as i32 - from.0 as i32;
        let dy = to.1 as i32 - from.1 as i32;
        
        // Use Bresenham's line algorithm for proper line drawing
        let steps = dx.abs().max(dy.abs());
        
        if steps == 0 {
            return;
        }
        
        let x_step = dx as f32 / steps as f32;
        let y_step = dy as f32 / steps as f32;
        
        for i in 0..=steps {
            let x = (from.0 as f32 + i as f32 * x_step) as u16;
            let y = (from.1 as f32 + i as f32 * y_step) as u16;
            
            if x < area.x + area.width && y < area.y + area.height {
                let point_area = Rect {
                    x,
                    y,
                    width: 1,
                    height: 1,
                };
                
                // Choose line character based on direction
                let line_char = if dx.abs() > dy.abs() {
                    if dx > 0 { "─" } else { "─" }
                } else if dy.abs() > dx.abs() {
                    if dy > 0 { "│" } else { "│" }
                } else {
                    // Diagonal
                    if (dx > 0 && dy > 0) || (dx < 0 && dy < 0) { "╲" } else { "╱" }
                };
                
                let line_widget = Paragraph::new(line_char)
                    .style(Style::default().fg(Color::Red));
                
                if point_area.intersects(area) {
                    f.render_widget(line_widget, point_area);
                }
            }
        }
        
        // Draw arrow heads to show direction (from FK to PK)
        self.draw_arrow_head(f, to, from, area);
    }
    
    fn draw_arrow_head(
        &self,
        f: &mut Frame,
        tip: (u16, u16),
        from: (u16, u16),
        area: Rect,
    ) {
        let dx = from.0 as i32 - tip.0 as i32;
        let dy = from.1 as i32 - tip.1 as i32;
        
        // Determine arrow character based on direction
        let arrow_char = if dx.abs() > dy.abs() {
            if dx > 0 { "◄" } else { "►" }
        } else {
            if dy > 0 { "▲" } else { "▼" }
        };
        
        let arrow_area = Rect {
            x: tip.0,
            y: tip.1,
            width: 1,
            height: 1,
        };
        
        let arrow_widget = Paragraph::new(arrow_char)
            .style(Style::default().fg(Color::Red).add_modifier(Modifier::BOLD));
        
        if arrow_area.intersects(area) {
            f.render_widget(arrow_widget, arrow_area);
        }
    }

    fn calculate_entity_area(&self, entity: &Entity, canvas_area: Rect) -> Rect {
        let x_ratio = entity.position.x / (self.canvas_width as f64);
        let y_ratio = entity.position.y / (self.canvas_height as f64);

        let x = (x_ratio * canvas_area.width as f64) as u16 + canvas_area.x;
        let y = (y_ratio * canvas_area.height as f64) as u16 + canvas_area.y;

        let width = entity.dimensions.width.min(canvas_area.width / 4);
        let height = entity.dimensions.height.min(canvas_area.height / 4);

        Rect { x, y, width, height }
    }
}

pub fn render_help_screen(f: &mut Frame, area: Rect) {
    let help_text = vec![
        Line::from("nERD - Terminal ERD Tool"),
        Line::from(""),
        Line::from("Navigation:"),
        Line::from("  q          - Quit application"),
        Line::from("  s          - Switch to SQL editor mode"),
        Line::from("  n          - Create new entity"),
        Line::from("  i          - Import SQL (or switch to editor)"),
        Line::from("  g          - Generate SQL from diagram"),
        Line::from("  r          - Refresh/re-layout diagram"),
        Line::from("  v          - Validate schema"),
        Line::from("  ?          - Show this help screen"),
        Line::from("  Esc        - Return to diagram view"),
        Line::from(""),
        Line::from("Entity Selection:"),
        Line::from("  Tab        - Select next entity"),
        Line::from("  Shift+Tab  - Select previous entity"),
        Line::from("  ↑↓←→       - Move selected entity"),
        Line::from("  Ctrl+D/Del - Delete selected entity"),
        Line::from(""),
        Line::from("SQL Editor:"),
        Line::from("  Ctrl+Enter - Sync SQL changes to diagram"),
        Line::from("  Type       - Enter/edit SQL statements"),
        Line::from("  g          - Generate SQL from current diagram"),
        Line::from(""),
        Line::from("Symbols:"),
        Line::from("  🗝         - Primary key column"),
        Line::from("  🔗         - Foreign key column"),
        Line::from("  ?          - Nullable column"),
        Line::from("  ─│╲╱►◄▲▼   - Red relationship lines with arrows"),
        Line::from(""),
        Line::from("Selected entities are highlighted in yellow."),
        Line::from("Red relationship lines connect exact columns (FK → PK)."),
        Line::from("Press any key to return."),
    ];

    let help = Paragraph::new(help_text)
        .block(
            Block::default()
                .title("Help")
                .title_style(Style::default().fg(Color::Cyan).add_modifier(Modifier::BOLD))
                .borders(Borders::ALL)
                .border_style(Style::default().fg(Color::Cyan)),
        )
        .style(Style::default().fg(Color::White));

    let centered_area = centered_rect(60, 80, area);
    f.render_widget(Clear, centered_area);
    f.render_widget(help, centered_area);
}

pub fn render_sql_editor(f: &mut Frame, content: &str, area: Rect) {
    let instructions = if content.is_empty() {
        "Enter SQL CREATE TABLE statements here.\nPress Ctrl+Enter to parse and apply.\nPress Esc to return to diagram view."
    } else {
        ""
    };
    
    let display_content = if content.is_empty() {
        instructions
    } else {
        content
    };

    let sql_text = Paragraph::new(display_content)
        .block(
            Block::default()
                .title("SQL Editor")
                .title_style(Style::default().fg(Color::Green).add_modifier(Modifier::BOLD))
                .borders(Borders::ALL)
                .border_style(Style::default().fg(Color::Green)),
        )
        .style(if content.is_empty() { 
            Style::default().fg(Color::DarkGray) 
        } else { 
            Style::default().fg(Color::White) 
        });

    f.render_widget(sql_text, area);
}

pub fn render_entity_creator(f: &mut Frame, buffer: &str, area: Rect) {
    let instructions = "Enter entity name and press Enter to create.\nPress Esc to cancel.";
    let display_text = if buffer.is_empty() {
        format!("{}\n\nEntity name: _", instructions)
    } else {
        format!("{}\n\nEntity name: {}_", instructions, buffer)
    };

    let text = Paragraph::new(display_text)
        .block(
            Block::default()
                .title("Create New Entity")
                .title_style(Style::default().fg(Color::Magenta).add_modifier(Modifier::BOLD))
                .borders(Borders::ALL)
                .border_style(Style::default().fg(Color::Magenta)),
        )
        .style(Style::default().fg(Color::White));

    let centered_area = centered_rect(50, 30, area);
    f.render_widget(Clear, centered_area);
    f.render_widget(text, centered_area);
}

fn centered_rect(percent_x: u16, percent_y: u16, r: Rect) -> Rect {
    let popup_layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints([
            Constraint::Percentage((100 - percent_y) / 2),
            Constraint::Percentage(percent_y),
            Constraint::Percentage((100 - percent_y) / 2),
        ])
        .split(r);

    Layout::default()
        .direction(Direction::Horizontal)
        .constraints([
            Constraint::Percentage((100 - percent_x) / 2),
            Constraint::Percentage(percent_x),
            Constraint::Percentage((100 - percent_x) / 2),
        ])
        .split(popup_layout[1])[1]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{Dimensions, Position};

    #[test]
    fn test_calculate_entity_area() {
        let renderer = DiagramRenderer::new(800, 600);
        
        let entity = Entity {
            name: "test".to_string(),
            columns: vec![],
            position: Position { x: 400.0, y: 300.0 },
            dimensions: Dimensions { width: 20, height: 10 },
        };

        let canvas_area = Rect {
            x: 0,
            y: 0,
            width: 100,
            height: 50,
        };

        let entity_area = renderer.calculate_entity_area(&entity, canvas_area);
        
        assert_eq!(entity_area.x, 50);  // 400/800 * 100
        assert_eq!(entity_area.y, 25);  // 300/600 * 50
        assert_eq!(entity_area.width, 20);
        assert_eq!(entity_area.height, 10);
    }

    #[test]
    fn test_centered_rect() {
        let area = Rect {
            x: 0,
            y: 0,
            width: 100,
            height: 50,
        };

        let centered = centered_rect(50, 60, area);
        
        assert_eq!(centered.width, 50);
        assert_eq!(centered.height, 30);
        assert_eq!(centered.x, 25);
        assert_eq!(centered.y, 10);
    }

    #[test]
    fn test_calculate_column_y_position() {
        use crate::models::Column;
        
        let renderer = DiagramRenderer::new(800, 600);
        
        let entity = Entity {
            name: "test_table".to_string(),
            columns: vec![
                Column {
                    name: "id".to_string(),
                    data_type: "INT".to_string(),
                    nullable: false,
                    is_primary_key: true,
                    is_foreign_key: false,
                    references: None,
                },
                Column {
                    name: "user_id".to_string(),
                    data_type: "INT".to_string(),
                    nullable: false,
                    is_primary_key: false,
                    is_foreign_key: true,
                    references: None,
                },
                Column {
                    name: "name".to_string(),
                    data_type: "VARCHAR(100)".to_string(),
                    nullable: false,
                    is_primary_key: false,
                    is_foreign_key: false,
                    references: None,
                },
            ],
            position: Position { x: 100.0, y: 100.0 },
            dimensions: Dimensions { width: 30, height: 10 },
        };

        let entity_area = Rect {
            x: 10,
            y: 10,
            width: 30,
            height: 10,
        };

        // Test first column (id) - should be at y=11 (10 + 1 for border + 0 for index)
        let id_y = renderer.calculate_column_y_position(&entity, "id", &entity_area);
        assert_eq!(id_y, 11);

        // Test second column (user_id) - should be at y=12 (10 + 1 for border + 1 for index)
        let user_id_y = renderer.calculate_column_y_position(&entity, "user_id", &entity_area);
        assert_eq!(user_id_y, 12);

        // Test third column (name) - should be at y=13 (10 + 1 for border + 2 for index)
        let name_y = renderer.calculate_column_y_position(&entity, "name", &entity_area);
        assert_eq!(name_y, 13);

        // Test non-existent column - should default to first column position
        let unknown_y = renderer.calculate_column_y_position(&entity, "unknown", &entity_area);
        assert_eq!(unknown_y, 11);
    }
}