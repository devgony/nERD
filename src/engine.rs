use std::fmt::Display;

use gluesql_core::{ast::DataType, data::Schema};

#[derive(PartialEq, Debug)]
pub struct Entity {
    pub name: String,
    pub attributes: Vec<Attribute>,
    pub x: usize,
    pub y: usize,
}

#[derive(PartialEq, Debug)]
pub struct Attribute {
    pub name: String,
    pub data_type: DataType,
    pub reffered_by: Option<(String, String)>,
    pub reffering_to: Option<(String, String)>,
}

pub fn into_entities(schemas: Vec<Schema>) -> Vec<Entity> {
    const LAYOUT_SIZE: usize = 100;
    const MARGIN: usize = 1;
    const ENTITY_WIDTH: usize = 12; // includes line width
    const GAP: usize = 3;
    // let mut max_y = MARGIN;

    schemas
        .iter()
        .enumerate()
        .fold(Vec::new(), |mut acc, (index, schema)| {
            let attributes: Vec<Attribute> = schema
                .column_defs
                .as_ref()
                .map(|column_defs| {
                    column_defs
                        .iter()
                        .map(|column_def| {
                            let reffering_to = schema.foreign_keys.iter().find_map(|fk| {
                                (fk.referencing_column_name == column_def.name).then_some((
                                    fk.referenced_table_name.clone(),
                                    fk.referenced_column_name.clone(),
                                ))
                            });
                            let reffered_by = schemas.iter().find_map(|sch| {
                                sch.foreign_keys
                                    .iter()
                                    .find(|fk| {
                                        fk.referenced_table_name == schema.table_name
                                            && fk.referenced_column_name == column_def.name
                                    })
                                    .map(|fk| {
                                        (sch.table_name.clone(), fk.referencing_column_name.clone())
                                    })
                            });

                            Attribute {
                                name: column_def.name.clone(),
                                data_type: column_def.data_type.clone(),
                                reffered_by,
                                reffering_to,
                            }
                        })
                        .collect()
                })
                .unwrap_or_default();

            // width of layout: 100
            // height of layout: 100
            // init starts with 1,1
            // width of Entity: 12
            // height of Entity: 1 + 1 + 1 + columns.len() + 1
            // gap between Entity: 3
            // if x is over than 400, move next line, which means it should start over the max(y)
            // should memorize max(y) with scan? => temporarily use mut global var

            let number_of_entities_in_row =
                ((LAYOUT_SIZE - 2 * MARGIN) as f64 / (ENTITY_WIDTH + GAP) as f64).floor() as usize; // should subtract 1 gap at the right most

            let x = MARGIN + (index % number_of_entities_in_row) * (ENTITY_WIDTH + GAP);
            let y = acc
                .iter()
                .rfind(|entity| entity.x == x)
                .map(|entity| entity.y + entity.attributes.len() + 4 + GAP)
                .unwrap_or(MARGIN);

            // let y =
            // let y = max_y;
            // let (x, y) = if x > LAYOUT_SIZE {
            //     (
            //         MARGIN
            //             + (ENTITY_WIDTH + GAP)
            //                 * (index % ((LAYOUT_SIZE - MARGIN) / (ENTITY_WIDTH + GAP))),
            //         max_y + attributes.len() + 3 + GAP,
            //     )
            // } else {
            //     (x, y)
            // };

            // if x > LAYOUT_SIZE {
            //     max_y = y;
            // }

            acc.push(Entity {
                name: schema.table_name.clone(),
                attributes,
                x,
                y,
            });

            acc
        })
}

#[cfg(test)]
mod tests {
    use gluesql_core::data::Schema;

    use super::*;

    #[test]
    fn test_engine() {
        let sqls = "
CREATE TABLE Tab1 (col1 INT, col2 INT);
CREATE TABLE Tab2 (col1 INT, col2 INT);
CREATE TABLE Tab3 (col1 INT, col2 INT);
CREATE TABLE Tab4 (col1 INT, col2 INT);
CREATE TABLE Tab5 (col1 INT, col2 INT);
CREATE TABLE Tab6 (col1 INT, col2 INT);
CREATE TABLE Tab7 (col1 INT, col2 INT);
CREATE TABLE Tab8 (col1 INT, col2 INT);
CREATE TABLE Tab9 (col1 INT, col2 INT);
CREATE TABLE Tab10 (col1 INT, col2 INT);
";
        let schemas = sqls
            .split(";")
            .filter_map(|sql| Schema::from_ddl(sql).ok())
            .collect::<Vec<_>>();

        let entities = into_entities(schemas);

        let positions: Vec<(String, usize, usize)> = entities
            .iter()
            .map(|entity| (entity.name.clone(), entity.x, entity.y))
            .collect();

        assert_eq!(
            positions,
            vec![
                ("Tab1".to_owned(), 1, 1),
                ("Tab2".to_owned(), 16, 1),
                ("Tab3".to_owned(), 31, 1),
                ("Tab4".to_owned(), 46, 1),
                ("Tab5".to_owned(), 61, 1),
                ("Tab6".to_owned(), 76, 1),
                ("Tab7".to_owned(), 91, 1),
                ("Tab8".to_owned(), 1, 9),
                ("Tab9".to_owned(), 16, 9),
                ("Tab10".to_owned(), 31, 9)
            ]
        );
    }
}
