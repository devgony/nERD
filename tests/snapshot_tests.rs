use gluesql_core::data::Schema;
use nERD::engine::into_entities;
use nERD::render::{render, render_foreign_key};
use insta::assert_snapshot;

#[test]
fn test_simple_table_rendering() {
    let sql = r#"
        CREATE TABLE users (
            id INT PRIMARY KEY,
            email TEXT,
            name TEXT
        );
    "#;
    
    let schemas: Vec<Schema> = sql
        .split(";")
        .map(|sql| sql.trim())
        .filter(|sql| !sql.is_empty())
        .filter_map(|sql| Schema::from_ddl(sql).ok())
        .collect();
    
    let entities = into_entities(schemas);
    
    let canvas_width = 100;
    let canvas_height = 50;
    let mut canvas: Vec<Vec<char>> = vec![vec![' '; canvas_width]; canvas_height];
    
    let (_, rendered_string) = render(canvas, &entities);
    
    assert_snapshot!(rendered_string);
}

#[test]
fn test_multiple_tables_rendering() {
    let sql = r#"
        CREATE TABLE users (
            id INT PRIMARY KEY,
            email TEXT,
            name TEXT,
            created_at TIMESTAMP
        );
        
        CREATE TABLE posts (
            id INT PRIMARY KEY,
            title TEXT,
            content TEXT,
            user_id INT,
            FOREIGN KEY (user_id) REFERENCES users(id)
        );
    "#;
    
    let schemas: Vec<Schema> = sql
        .split(";")
        .map(|sql| sql.trim())
        .filter(|sql| !sql.is_empty())
        .filter_map(|sql| Schema::from_ddl(sql).ok())
        .collect();
    
    let entities = into_entities(schemas);
    
    let canvas_width = 100;
    let canvas_height = 50;
    let mut canvas: Vec<Vec<char>> = vec![vec![' '; canvas_width]; canvas_height];
    
    let (_, rendered_string) = render(canvas, &entities);
    
    assert_snapshot!(rendered_string);
}

#[test]
fn test_foreign_key_rendering() {
    let sql = r#"
        CREATE TABLE users (
            id INT PRIMARY KEY,
            email TEXT,
            name TEXT
        );
        
        CREATE TABLE posts (
            id INT PRIMARY KEY,
            title TEXT,
            user_id INT,
            FOREIGN KEY (user_id) REFERENCES users(id)
        );
        
        CREATE TABLE comments (
            id INT PRIMARY KEY,
            content TEXT,
            post_id INT,
            user_id INT,
            FOREIGN KEY (post_id) REFERENCES posts(id),
            FOREIGN KEY (user_id) REFERENCES users(id)
        );
    "#;
    
    let schemas: Vec<Schema> = sql
        .split(";")
        .map(|sql| sql.trim())
        .filter(|sql| !sql.is_empty())
        .filter_map(|sql| Schema::from_ddl(sql).ok())
        .collect();
    
    let entities = into_entities(schemas);
    
    let canvas_width = 100;
    let canvas_height = 50;
    let mut canvas: Vec<Vec<char>> = vec![vec![' '; canvas_width]; canvas_height];
    
    let (canvas_with_entities, _) = render(canvas, &entities);
    let (_, rendered_with_fks) = render_foreign_key(canvas_with_entities, &entities);
    
    assert_snapshot!(rendered_with_fks);
}

#[test]
fn test_self_referencing_table() {
    let sql = r#"
        CREATE TABLE categories (
            id INT PRIMARY KEY,
            name TEXT,
            parent_id INT,
            FOREIGN KEY (parent_id) REFERENCES categories(id)
        );
    "#;
    
    let schemas: Vec<Schema> = sql
        .split(";")
        .map(|sql| sql.trim())
        .filter(|sql| !sql.is_empty())
        .filter_map(|sql| Schema::from_ddl(sql).ok())
        .collect();
    
    let entities = into_entities(schemas);
    
    let canvas_width = 100;
    let canvas_height = 50;
    let mut canvas: Vec<Vec<char>> = vec![vec![' '; canvas_width]; canvas_height];
    
    let (canvas_with_entities, _) = render(canvas, &entities);
    let (_, rendered_with_fks) = render_foreign_key(canvas_with_entities, &entities);
    
    assert_snapshot!(rendered_with_fks);
}

#[test]
fn test_complex_schema_rendering() {
    let sql = r#"
        CREATE TABLE users (
            id INT PRIMARY KEY,
            email TEXT,
            name TEXT,
            created_at TIMESTAMP
        );
        
        CREATE TABLE categories (
            id INT PRIMARY KEY,
            name TEXT,
            parent_id INT,
            FOREIGN KEY (parent_id) REFERENCES categories(id)
        );
        
        CREATE TABLE products (
            id INT PRIMARY KEY,
            name TEXT,
            category_id INT,
            price INT,
            stock_quantity INT,
            created_by INT,
            FOREIGN KEY (category_id) REFERENCES categories(id),
            FOREIGN KEY (created_by) REFERENCES users(id)
        );
        
        CREATE TABLE orders (
            id INT PRIMARY KEY,
            user_id INT,
            status TEXT,
            total_amount INT,
            created_at TIMESTAMP,
            FOREIGN KEY (user_id) REFERENCES users(id)
        );
        
        CREATE TABLE order_items (
            id INT PRIMARY KEY,
            order_id INT,
            product_id INT,
            quantity INT,
            price INT,
            FOREIGN KEY (order_id) REFERENCES orders(id),
            FOREIGN KEY (product_id) REFERENCES products(id)
        );
    "#;
    
    let schemas: Vec<Schema> = sql
        .split(";")
        .map(|sql| sql.trim())
        .filter(|sql| !sql.is_empty())
        .filter_map(|sql| Schema::from_ddl(sql).ok())
        .collect();
    
    let entities = into_entities(schemas);
    
    let canvas_width = 150;
    let canvas_height = 80;
    let mut canvas: Vec<Vec<char>> = vec![vec![' '; canvas_width]; canvas_height];
    
    let (canvas_with_entities, _) = render(canvas, &entities);
    let (_, rendered_with_fks) = render_foreign_key(canvas_with_entities, &entities);
    
    assert_snapshot!(rendered_with_fks);
}

#[test]
fn test_table_with_many_columns() {
    let sql = r#"
        CREATE TABLE detailed_users (
            id INT PRIMARY KEY,
            username TEXT,
            email TEXT,
            first_name TEXT,
            last_name TEXT,
            phone TEXT,
            address TEXT,
            city TEXT,
            country TEXT,
            postal_code TEXT,
            date_of_birth TIMESTAMP,
            created_at TIMESTAMP,
            updated_at TIMESTAMP,
            is_active BOOLEAN,
            last_login TIMESTAMP
        );
    "#;
    
    let schemas: Vec<Schema> = sql
        .split(";")
        .map(|sql| sql.trim())
        .filter(|sql| !sql.is_empty())
        .filter_map(|sql| Schema::from_ddl(sql).ok())
        .collect();
    
    let entities = into_entities(schemas);
    
    let canvas_width = 100;
    let canvas_height = 50;
    let mut canvas: Vec<Vec<char>> = vec![vec![' '; canvas_width]; canvas_height];
    
    let (_, rendered_string) = render(canvas, &entities);
    
    assert_snapshot!(rendered_string);
}