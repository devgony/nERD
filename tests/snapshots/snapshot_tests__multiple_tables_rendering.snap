---
source: tests/snapshot_tests.rs
expression: rendered_string
---
                                                                                                    
 ┌──────────┐   ┌──────────┐                                                                        
 │users     │   │posts     │                                                                        
 ├──────────┤   ├──────────┤                                                                        
 │id        │   │id        │                                                                        
 │email     │   │title     │                                                                        
 │name      │   │content   │                                                                        
 │created_at│   │user_id   │                                                                        
 └──────────┘   └──────────┘
